desc/coils.py

```diff
@@ -0,0 +1,457 @@
+        for i in range(n):
+            coili = coil.copy()
+            coili.rotate(axis, angle=phis[i])
+            coili.current = currents[i]
+            coils.append(coili)
+        return cls(*coils)
+
+    @classmethod
+    def linspaced_linear(
+        cls, coil, current=None, displacement=[2, 0, 0], n=4, endpoint=False
+    ):
+        """Create a coil group by repeating a coil n times in a straight line.
+
+        Parameters
+        ----------
+        coil : Coil
+            base coil to repeat
+        current : float or array-like, shape(n,)
+            current in (each) coil
+        displacement : array-like, shape(3,)
+            total displacement of the final coil
+        n : int
+            number of copies of original coil
+        endpoint : bool
+            whether to include a coil at final point
+        """
+        assert isinstance(coil, Coil)
+        if current is None:
+            current = coil.current
+        currents = jnp.broadcast_to(current, (n,))
+        displacement = jnp.asarray(displacement)
+        coils = []
+        a = jnp.linspace(0, 1, n, endpoint=endpoint)
+        for i in range(n):
+            coili = coil.copy()
+            coili.translate(a[i] * displacement)
+            coili.current = currents[i]
+            coils.append(coili)
+        return cls(*coils)
+
+    @classmethod
+    def from_symmetry(cls, coils, NFP, sym=False):
+        """Create a coil group by reflection and symmetry
+
+        Given coils over one field period, repeat coils NFP times between
+        0 and 2pi to form full coil set.
+
+        Or, give coils over 1/2 of a field period, repeat coils 2*NFP times
+        between 0 and 2pi to form full stellarator symmetric coil set.
+
+        Parameters
+        ----------
+        coils : Coil, CoilGroup, Coilset
+            base coil or collection of coils to repeat
+        NFP : int
+            number of field periods
+        sym : bool
+            whether coils should be stellarator symmetric
+        """
+        if not isinstance(coils, CoilSet):
+            coils = CoilSet(coils)
+        coilset = []
+        if sym:
+            # first reflect/flip original coilset
+            # ie, given coils [1,2,3] at angles [0, pi/6, 2pi/6]
+            # we want a new set like [1,2,3,flip(3),flip(2),flip(1)]
+            # at [0, pi/6, 2pi/6, 3pi/6, 4pi/6, 5pi/6]
+            flipped_coils = []
+            normal = jnp.array([-jnp.sin(jnp.pi / NFP), jnp.cos(jnp.pi / NFP), 0])
+            for coil in coils[::-1]:
+                fcoil = coil.copy()
+                fcoil.flip(normal)
+                fcoil.flip([0, 0, 1])
+                fcoil.current = -1 * coil.current
+                flipped_coils.append(fcoil)
+            coils = coils + flipped_coils
+        for k in range(0, NFP):
+            coil = coils.copy()
+            coil.rotate(axis=[0, 0, 1], angle=2 * jnp.pi * k / NFP)
```

Comment:
Won't the `k=0` iteration of this just duplicate the existing field period of coils without rotating them?

Reply:
never mind